File: samples/AndroidSamples/salmon-android-samples/src/main/java/com/mku/salmon/samples/main/HttpDriveActivity.java
package com.mku.salmon.samples.main;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.textfield.TextInputEditText;
import com.mku.android.fs.file.AndroidFileSystem;
import com.mku.fs.file.HttpFile;
import com.mku.fs.file.IFile;
import com.mku.salmon.samples.R;
import com.mku.salmon.samples.samples.DriveSample;
import com.mku.salmon.samples.utils.AndroidFileChooser;
import com.mku.salmon.streams.AesStream;
import com.mku.salmon.streams.ProviderType;
import com.mku.salmonfs.drive.AesDrive;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class HttpDriveActivity extends AppCompatActivity {
    public static final int REQUEST_EXPORT_FILES = 1003;
    private EditText httpURL;
    private TextInputEditText password;
    private Button openDriveButton;
    private Button listFilesButton;
    private Button exportFilesButton;
    private Button closeDriveButton;
    private EditText outputText;

    private int threads = 1;
    private static final String defaultPassword = "test123";
    String defaultHttpDriveURL = "";
    private AesDrive httpDrive;

    private final Executor executor = Executors.newCachedThreadPool();

    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);

        setContentView(R.layout.activity_http_drive);

        httpURL = findViewById(R.id.HTTP_DRIVE_LOCATION);
        httpURL.setText(defaultHttpDriveURL);
        password = findViewById(R.id.TEXT_PASSWORD);

        openDriveButton = findViewById(R.id.OPEN_DRIVE_BUTTON);
        openDriveButton.setOnClickListener((e) -> {
            executor.execute(() -> {
                openDrive();
            });
        });

        listFilesButton = findViewById(R.id.LIST_DRIVE_BUTTON);
        listFilesButton.setOnClickListener((e) -> {
            executor.execute(() -> {
                listFiles();
            });
        });
        exportFilesButton = findViewById(R.id.EXPORT_FILES_BUTTON);
        exportFilesButton.setOnClickListener((e) -> {
            AndroidFileChooser.openFilesystem(this, REQUEST_EXPORT_FILES, true);
        });

        closeDriveButton = findViewById(R.id.CLOSE_DRIVE_BUTTON);
        closeDriveButton.setOnClickListener((e) -> {
            executor.execute(() -> {
                closeDrive();
            });
        });

        outputText = findViewById(R.id.OUTPUT_TEXT);
        password.setText(defaultPassword);

        AndroidFileSystem.initialize(this);
        AesStream.setAesProviderType(ProviderType.Default);

        initialize();
    }

    public void log(String msg) {
        runOnUiThread(() -> {
            outputText.append(msg + "\n");
        });
    }

    public void clearLog() {
        runOnUiThread(() -> {
            outputText.setText("");
        });
    }

    private void initialize() {
        AndroidFileSystem.initialize(this);
        AesStream.setAesProviderType(ProviderType.Default);
    }

    public void openDrive() {
        clearLog();

        try {
            IFile driveDir = new HttpFile(httpURL.getText().toString());
            httpDrive = DriveSample.openDrive(driveDir, password.getText().toString(),
                    this::log);
        } catch (Exception e) {
            e.printStackTrace();
            log(e.getMessage());
        }
    }

    public void listFiles() {
        try {
            DriveSample.listFiles(httpDrive, this::log);
        } catch (Exception e) {
            e.printStackTrace();
            log(e.getMessage());
        }
    }

    public void exportFiles(IFile exportDir) {
        try {
            DriveSample.exportFiles(httpDrive, exportDir, threads, this::log);
        } catch (Exception e) {
            e.printStackTrace();
            log(e.getMessage());
        }
    }

    public void closeDrive() {
        DriveSample.closeDrive(httpDrive, this::log);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (data == null)
            return;
        android.net.Uri uri = data.getData();
        AndroidFileChooser.setUriPermissions(this, data, uri);
        if (requestCode == REQUEST_EXPORT_FILES) {
            IFile exportDir = AndroidFileSystem.getRealFile(uri.toString(), true);
            executor.execute(() -> {
                exportFiles(exportDir);
            });
        }
    }
}

File: samples/DotNetAndroidSamples/DotNetAndroidSamples/Main/HttpDriveActivity.cs
using AndroidX.AppCompat.App;
using Android.Content;
using Android.Widget;
using Android.OS;
using System;
using DotNetAndroidSamples;
using Google.Android.Material.TextField;
using Mku.Android.FS.File;
using Mku.FS.File;
using Mku.Salmon.Samples.Samples;
using Mku.Salmon.Streams;
using Mku.SalmonFS.Drive;
using Mku.Salmon.Samples.Utils;
using System.Threading.Tasks;
using Android.App;
using Uri = Android.Net.Uri;
using Mku.Salmon.Transform;
using Mku.Android.Salmon.Transform;

namespace Mku.Salmon.Samples.Main;

[Activity(Label = "@string/app_name", Theme = "@style/AppTheme",
    ScreenOrientation = global::Android.Content.PM.ScreenOrientation.Portrait)]
public class HttpDriveActivity : AppCompatActivity
{
    public const int REQUEST_EXPORT_FILES = 1003;
    private EditText httpURL;
    private TextInputEditText password;
    private Button openDriveButton;
    private Button listFilesButton;
    private Button exportFilesButton;
    private Button closeDriveButton;
    private EditText outputText;

    private int threads = 1;
    private static readonly string defaultPassword = "test";
    string defaultHttpDriveURL = "https://localhost/testvault";
    string httpUser = "user";
    string httpPassword = "password";

    private AesDrive httpDrive;

    protected override void OnCreate(Bundle bundle)
    {
        base.OnCreate(bundle);

        // enable only if you're testing with an HTTP server
		// In all other cases you should be using an HTTPS server
        // HttpSyncClient.AllowClearTextTraffic = true;

        SetContentView(Resource.Layout.activity_http_drive);

        httpURL = (EditText)FindViewById(Resource.Id.HTTP_DRIVE_LOCATION);
        httpURL.Text = defaultHttpDriveURL;
        password = (TextInputEditText)FindViewById(Resource.Id.TEXT_PASSWORD);

        openDriveButton = (Button)FindViewById(Resource.Id.OPEN_DRIVE_BUTTON);
        openDriveButton.Click += (s, e) =>
        {
            Task.Run(() =>
            {
                OpenDrive();
            });
        };

        listFilesButton = (Button)FindViewById(Resource.Id.LIST_DRIVE_BUTTON);
        listFilesButton.Click += (s, e) =>
        {
            Task.Run(() =>
            {
                ListFiles();
            });
        };
        exportFilesButton = (Button)FindViewById(Resource.Id.EXPORT_FILES_BUTTON);
        exportFilesButton.Click += (s, e) =>
        {
            AndroidFileChooser.OpenFilesystem(this, REQUEST_EXPORT_FILES, true);
        };

        closeDriveButton = (Button)FindViewById(Resource.Id.CLOSE_DRIVE_BUTTON);
        closeDriveButton.Click += (s, e) =>
        {
            Task.Run(() =>
            {
                CloseDrive();
            });
        };

        outputText = (EditText)FindViewById(Resource.Id.OUTPUT_TEXT);
        password.Text = defaultPassword;

        AesStream.AesProviderType = ProviderType.Default;

        Initialize();
    }

    public void Log(string msg)
    {
        RunOnUiThread(() =>
        {
            outputText.Append(msg + "\n");
        });
    }

    public void ClearLog()
    {
        RunOnUiThread(() =>
        {
            outputText.Text = "";
        });
    }

    private void Initialize()
    {
		AesNativeTransformer.NativeProxy = new AndroidNativeProxy();
        AndroidFileSystem.Initialize(this);
        AesStream.AesProviderType = ProviderType.Default;
    }

    public void OpenDrive()
    {
        ClearLog();
        try
        {
            IFile driveDir = new HttpFile(httpURL.Text.ToString(), new Credentials(httpUser, httpPassword));
            httpDrive = DriveSample.OpenDrive(driveDir, password.Text.ToString(),
                    Log);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine(e);
            Log(e.Message);
        }
    }

    public void ListFiles()
    {
        try
        {
            DriveSample.ListFiles(httpDrive, Log);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine(e);
            Log(e.Message);
        }
    }

    public void ExportFiles(IFile exportDir)
    {

        try
        {
            DriveSample.ExportFiles(httpDrive, exportDir, threads, Log);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine(e);
            Log(e.Message);
        }
    }

    public void CloseDrive()
    {
        DriveSample.CloseDrive(httpDrive, Log);
    }

    protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
    {
        base.OnActivityResult(requestCode, resultCode, data);
        if (data == null)
            return;
        Uri uri = data.Data;
        AndroidFileChooser.SetUriPermissions(this, data, uri);
        if (requestCode == REQUEST_EXPORT_FILES)
        {
            IFile exportDir = AndroidFileSystem.GetRealFile(uri.ToString(), true);
            Task.Run(() =>
            {
                ExportFiles(exportDir);
            });
        }
    }
}